alts: Major rewrite of the ALTS transport socket to fix concurrent connection establishment bugs (version 2).

matthewstevenson88 opened the pull request:
Commit Message:
Additional Description: The existing ALTS transport socket extension cannot handle establishing multiple concurrent connections. This is due to a fundamental problem in how the transport socket is architected on top of the ALTS TSI implementation in the gRPC C-core; specifically, this issue occurs because the ALTS transport socket uses gRPC-internal APIs to try to replicate the gRPC event queue, but fails to do so correctly. To avoid this problem and problems like it in the future, this PR rewrites the ALTS transport socket so that it only depends on public gRPC APIs. We have confirmed that this fixes the original bug and the transport socket can now successfully establish multiple concurrent connections.
Risk Level:
Testing:
Docs Changes:
Release Notes: Rewrite of the ALTS transport socket to fix concurrent connection establishment bugs. The ALTS transport socket can now successfully establish multiple concurrent connections.
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

botengyao commented:
Hi @alyssawilk, feel free to assign the ALTS review to me. @matthewstevenson88, we can sync up offline if you need some help. From my last check, we need to fix some existing tests.

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

botengyao commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -334,9 +315,7 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
-    ASSERT(action == Network::PostIoAction::KeepOpen);
-    // TODO(lizan): Handle synchronous handshake when TsiHandshaker supports it.
-    return {Network::PostIoAction::KeepOpen, 0, false};
+    return {action, 0, action == Network::PostIoAction::Close ? true : false};
```
besides, is there an integration test cover this?

matthewstevenson88 replied:
Fixed.

AFAIK, there is no integration test in OSS Envoy that caught this. If you can point me to an example to base of, I'm happy to add one (or maybe we should defer this to a follow-up PR since this is an existing issue).

botengyao replied:
I am thinking like we should trigger the assertion if this line is triggered. We can leverage https://github.com/envoyproxy/envoy/blob/47afc8c884ca95a40d01513cf82f0f40b67b8ed8/test/extensions/transport_sockets/alts/alts_integration_test.cc#L195
to run into this condition during `doWrite` call.

matthewstevenson88 replied:
Sorry I missed this. There are tests in tsi_socket_test.cc that check this code block now.

botengyao replied:
Thanks!

botengyao commented on source/extensions/transport_sockets/alts/alts_channel_pool.cc:
```diff
@@ -0,0 +1,51 @@
+std::unique_ptr<AltsChannelPool>
+AltsChannelPool::create(absl::string_view handshaker_service_address) {
+  // TODO(matthewstevenson88): Extend this to be configurable through API.
+  std::size_t channel_pool_size = 10;
```
Maybe move this to be a const static number right now.

botengyao commented on source/extensions/transport_sockets/alts/alts_proxy.cc:
```diff
@@ -0,0 +1,151 @@
+}
+
+absl::StatusOr<HandshakerResp> AltsProxy::sendStartClientHandshakeReq() {
+  // Prepare the StartClientHandshakeReq message. Ignore the target name field:
```
maybe s/:/,

matthewstevenson88 replied:
Done.

botengyao reviewed (commented):
Thanks for working on this, and this looks great!
I haven't took a detailed look for the test, but overall the logic LGTM.

botengyao commented:
Besides, feel free to remove the `[WIP]` marker. and @envoyproxy/envoy-maintainers for the CI approval. Thanks!

jmarantz commented:
/retest

jmarantz commented:
@matthewstevenson88 if this is ready for final review please remove "[WIP]" from the PR title. Thanlk you!

/wait

matthewstevenson88 commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:
```diff
@@ -0,0 +1,61 @@
+namespace TransportSockets {
+namespace Alts {
+
+constexpr std::size_t kAltsMinFrameSize = 16 * 1024;
```
It is, yes.

matthewstevenson88 commented on source/extensions/transport_sockets/alts/config.cc:
```diff
@@ -63,81 +71,43 @@ createHandshakeValidator(const envoy::extensions::transport_sockets::alts::v3::A
+    std::unique_ptr<AltsTsiHandshaker> tsi_handshaker = nullptr;
     if (is_upstream) {
-      options = GrpcAltsCredentialsOptionsPtr(grpc_alts_credentials_client_options_create());
+      tsi_handshaker = AltsTsiHandshaker::createForClient(alts_shared_state->getChannel());
```
Good observation! We have the nullity check where the channel is ultimately used, in the `AltsProxy::create` method. We could also add nullity checks at each stage of the plumbing from here to there, but I think having just the one check should be fine. WDYT?

matthewstevenson88 commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -334,9 +315,7 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
     // dedicated thread in a separate API within which handshake_complete_
```
Good call, removed this since it is no longer relevant.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_channel_pool.cc:
```diff
@@ -0,0 +1,52 @@
+namespace Alts {
+
+// TODO(matthewstevenson88): Extend this to be configurable through API.
+constexpr std::size_t kChannelPoolSize = 10;
```
likewise

KBaichoo commented on source/extensions/transport_sockets/alts/alts_channel_pool.cc:
```diff
@@ -0,0 +1,52 @@
+
+// TODO(matthewstevenson88): Add logic to limit number of outstanding channels.
+std::shared_ptr<grpc::Channel> AltsChannelPool::getChannel() const {
+  absl::BitGen gen;
```
Creating a BitGen to dtor everytime seems expensive.  Do we need uniform randomness among channels, is round robin good enough? Can we use a SharedBit gen member?

matthewstevenson88 replied:
Good call, switched to round robin.

KBaichoo reviewed (commented):
Thank you for working on this. Here's a quick pass.

Can you add yourself as a codeowner https://github.com/envoyproxy/envoy/blob/620091ac80a3611f3340d21ff558666a9314d465/CODEOWNERS#L35 since this is a huge rewrite and you probably have the best expertise?

Can you add release notes? and add more details in the PR description.

Thanks!

/wait

matthewstevenson88 commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:
```diff
@@ -0,0 +1,61 @@
+struct AltsHandshakeResult {
+  TsiFrameProtectorPtr frame_protector;
+  std::string peer_identity;
+  std::string unused_bytes;
```
Switched to `std::vector<uint8_t>`.

matthewstevenson88 commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:
```diff
@@ -0,0 +1,61 @@
+};
+
+// Manages a single ALTS handshake.
+class AltsTsiHandshaker {
```
Before this PR, the ALTS transport socket relied on an implementation of the TSI interface from the gRPC C-core for the lower-level ALTS details. The tsi_handshaker was an Envoy wrapper over this interface from gRPC. All of the code in config.cc and tsi_socket.{h,cc} is built on top of tsi_handshaker, so I didn't want to entirely discard it in this PR (though I may do so in a follow-up with cleanup). Instead, I created alts_tsi_handshaker, which enables us to drop in the new functionality into the existing tsi_handshaker wrapper (with a few edits). Does that help clarify?

KBaichoo replied:
Thank you for the context here!

matthewstevenson88 commented:
> Thank you for working on this. Here's a quick pass.
> 
> Can you add yourself as a codeowner
> 
> https://github.com/envoyproxy/envoy/blob/620091ac80a3611f3340d21ff558666a9314d465/CODEOWNERS#L35
> 
> since this is a huge rewrite and you probably have the best expertise?
> Can you add release notes? and add more details in the PR description.
> 
> Thanks!

Thanks for the review, and thanks to @botengyao as well! I added myself as a code owner, added more details in the PR description, and added release notes in the PR description.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_channel_pool.h:
```diff
@@ -0,0 +1,39 @@
+  explicit AltsChannelPool(const std::vector<std::shared_ptr<grpc::Channel>>& channel_pool);
+
+  // Use round-robin to select channels from the pool.
+  absl::Mutex mu_;
```
can use `ABSL_GUARDED_BY` macros to annotate what this protects?

KBaichoo commented on CODEOWNERS:
```diff
@@ -32,7 +32,7 @@ extensions/filters/common/original_src @klarose @mattklein123
 /*/extensions/filters/http/grpc_http1_reverse_bridge @zuercher @mattklein123
 # alts transport socket extension
-/*/extensions/transport_sockets/alts @adisuissa @yangminzhu
+/*/extensions/transport_sockets/alts @adisuissa @yangminzhu @matthewstevenson88
```
I think you can remove @yangminzhu here.

KBaichoo commented on source/extensions/transport_sockets/alts/config.cc:
```diff
@@ -63,81 +71,43 @@ createHandshakeValidator(const envoy::extensions::transport_sockets::alts::v3::A
     ASSERT(local_address != nullptr);
-
-    GrpcAltsCredentialsOptionsPtr options;
+    std::unique_ptr<AltsTsiHandshaker> tsi_handshaker = nullptr;
```
no need to initialize nullptr as will do this in the default ctor.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.h:
```diff
@@ -0,0 +1,61 @@
+
+  // Exposed for testing purposes only.
+  absl::StatusOr<std::unique_ptr<AltsHandshakeResult>>
+  getHandshakeResult(const grpc::gcp::HandshakerResult& result, absl::string_view received_bytes,
```
That makes sense then to not call it for "ForTest". But  can you point to where we call it explicitly from a test e.g. such that it helps with coverage (I might have missed it). Otherwise if we just use it via the public interface in test, making it private sgtm.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  // Get a handshake message from the handshaker service.
+  absl::string_view in_bytes(reinterpret_cast<const char*>(received_bytes), received_bytes_size);
```
do you mean to drop the unsigned here?

matthewstevenson88 replied:
Fixed, replaced with `absl::Span<const uint8_t>`.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:
```diff
@@ -0,0 +1,169 @@
+      return alts_proxy.status();
+    }
+    alts_proxy_ = *std::move(alts_proxy);
+    if (is_client_) {
```
What do you think about splitting this out into a server / client subclasses which can share some common core but may differ in the implementation here (vs us doing work for both and mixing the logic together with if/else guards).

KBaichoo replied:
sg, thanks for your consideration.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_tsi_handshaker.cc:
```diff
@@ -0,0 +1,169 @@
+std::size_t AltsTsiHandshaker::computeMaxFrameSize(const grpc::gcp::HandshakerResult& result) {
+  std::size_t max_frame_size = AltsMinFrameSize;
+  if (result.max_frame_size() > 0) {
+    max_frame_size = std::min<std::size_t>(result.max_frame_size(), MaxFrameSize);
+    max_frame_size = std::max<std::size_t>(max_frame_size, AltsMinFrameSize);
+  }
```
std::clamp will do what you're doing here.

KBaichoo commented on test/extensions/transport_sockets/alts/alts_integration_test.cc:
```diff
@@ -87,6 +105,204 @@ class CapturingHandshakerService : public grpc::gcp::HandshakerService::Service
   size_t server_max_frame_size{0};
 };
 
+// FakeHandshakeService implements a fake handshaker service using a fake key
```
Do we an an integration test where we concurrently try to make connections (say 2) and we have 2 workers? I think that would help give confidence in fixing the underlying issue (and not introducing another race)

matthewstevenson88 replied:
I have an integration test where I run many concurrent connections through an Envoy running on GCP using the real ALTS handshaker service - this is what I used to confirm the fix.

At present, it is not clear to me how to set up a full integration test in the Envoy repo that does many concurrent connections with a fake ALTS handshaker service. However, I set up unit tests at the TSI handshaker layer that do concurrent handshakes, and these also would have caught the original issue.

KBaichoo commented on source/extensions/transport_sockets/alts/alts_proxy.h:
```diff
@@ -0,0 +1,68 @@
+  sendStartServerHandshakeReq(absl::Span<const uint8_t> in_bytes);
+
+  // Sends a NextHandshakeMessageReq message to the ALTS handshaker service and
+  // returns the response. This API is blocking.
```
What is being blocked here? e.g. are we blocking the Envoy silo here with any / all of the calls above? If so we can easily bring the system to a halt by the calls above.

matthewstevenson88 replied:
No sorry, I just mean that we're blocking the current thread. I removed this comment because I don't think it adds anything.

matthewstevenson88 replied:
@botengyao to correct me if I explain the TSI socket poorly. :)

When the ALTS transport socket is handling an ALTS handshake, this is occurring in a worker thread, and we are block this worker thread when we call this function and wait for this function to execute. Under the hood of this function, we are making a gRPC call to the ATLS handshaker service.

@KBaichoo Does that help clarify? Can I expand on anything?

Apologies I'm not sure what you mean by "the Envoy silo" above. If this does not answer your question, can you expand on this?

KBaichoo replied:
Thanks for clarifying, as such as can easily bring the entire proxy to a halt if the grpc service is unavailable / experiencing high latency. I think we should strongly document this, and possibly follow up. I suspect mostly Google might use this extension @yanavlasov @adisuissa @botengyao  on if/when this bottleneck would need to be addressed.

botengyao replied:
Right, we should document this. I think the best solution is to make this RPC call as an async stream with response callback, similar as the async gRPC client here. https://github.com/envoyproxy/envoy/blob/1140c93e207bc30995dfb5948bff9e4b48302f6f/source/common/grpc/async_client_impl.h#L12. Did we perform the load test with the current implementation internally with with currently logic?

matthewstevenson88 replied:
Thansk @KBaichoo for explaining, really glad you caught this. I've added documentation to alts_proxy.h and added a TODO in alts_proxy.cc to further investigate the async client.

@botengyao I have load testing set up with the internal implementation and I did not notice this issue there. For that reason, I'm tempted to suggest that we merge this PR (so that there is a working version of the ALTS transport socket) and look into a follow-up PR to use the async client. WDYT?

botengyao replied:
sgtm, defer to @KBaichoo for approval. Thanks!

botengyao replied:
Having saying that, and taking a deeper look, I think this still should be the **top priority TODO** to make the gRPC call be async to not block the worker thread based on Envoy's silo. I tend to let this PR merged since there is no actual usage and this makes the extension workable at least, but also glad to hear @KBaichoo and others' opinion.

KBaichoo replied:
+1 to letting this through since I expect there might not be significant usage of this (given the preexisting bug); this is a step in the right direction and this PR is already large enough.

KBaichoo commented on test/extensions/transport_sockets/alts/alts_integration_test.cc:
```diff
@@ -87,6 +105,204 @@ class CapturingHandshakerService : public grpc::gcp::HandshakerService::Service
+  // least STARTED. When the handshaker server produces the first fame, the
+  // state becomes SENT. After the handshaker server processes the final frame
+  // from the peer, the state becomes COMPLETED.
+  enum HandshakeState { INITIAL, STARTED, SENT, COMPLETED };
```
prefer enum class

matthewstevenson88 replied:
Done.

KBaichoo reviewed (commented):
Thanks for working on this, generally looks good apart from the "blocking" question

/wait

matthewstevenson88 commented:
> Thanks for working on this, generally looks good apart from the "blocking" question
> 
> /wait

Thanks for your review @KBaichoo - your comments/suggestions are much appreciated. :)

@botengyao Could you also take a final pass when you have a chance?

Thanks both!

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

botengyao commented on source/extensions/transport_sockets/alts/alts_proxy.cc:
```diff
@@ -0,0 +1,153 @@
+  if (handshaker_service_channel == nullptr) {
+    return absl::InvalidArgumentError("Handshaker service channel is null.");
+  }
+  auto client_context = std::make_unique<grpc::ClientContext>();
```
Do we need to set_deadline for the gRPC client through the ClientContext, since it is possible to end up waiting forever?

matthewstevenson88 replied:
Good idea, done.

botengyao reviewed (commented):
Thanks for working on this epic change! LGTM module the package and a question.

/wait

botengyao commented:
cc @phlax for CI. Thanks!

matthewstevenson88 commented on bazel/repository_locations.bzl:
```diff
@@ -386,6 +386,21 @@ REPOSITORY_LOCATIONS_SPEC = dict(
+        project_name = "Protobuf and gRPC rules for Bazel",
+        project_desc = "Bazel rules for building Protobuf and gRPC code and libraries from proto_library targets",
+        project_url = "https://github.com/rules-proto-grpc/rules_proto_grpc",
+        version = "4.4.0",
```
Done.

matthewstevenson88 commented on test/extensions/transport_sockets/alts/alts_proxy_test.cc:
```diff
@@ -0,0 +1,399 @@
+  std::unique_ptr<std::thread> server_thread_;
+};
+
+TEST_F(AltsProxyTest, ClientStartSuccess) {
```
Done.

matthewstevenson88 commented:
/retest

botengyao reviewed (commented):
LGTM, thanks!

@envoyproxy/dependency-shepherds or @RyanTheOptimist for dependency approval and CI.

matthewstevenson88 commented on source/extensions/transport_sockets/alts/alts_proxy.cc:
```diff
@@ -0,0 +1,157 @@
+using ::grpc::gcp::StartClientHandshakeReq;
+using ::grpc::gcp::StartServerHandshakeReq;
+
+constexpr absl::Duration AltsClientContextDeadline = absl::Seconds(30);
```
Done.

adisuissa commented:
@KBaichoo any more comments?

KBaichoo reviewed (commented):
Mostly looks good apart from conversation in https://github.com/envoyproxy/envoy/pull/29803#discussion_r1392930294

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

matthewstevenson88 commented:
/retest

adisuissa commented:
This is an extension-only modification, so I think !Google review is not required.
Assigning Alyssa for a senior-maintainer stamp.
/assign @alyssawilk
